drivers/android/process.rs

```diff
@@ -712,14 +703,15 @@ impl Process {
             return Ok(());
         }
 
-        let death = {
-            let mut pinned = Pin::from(death.write(
-                // SAFETY: `init` is called below.
-                unsafe { NodeDeath::new(info.node_ref.node.clone(), self.clone(), cookie) },
-            ));
-            pinned.as_mut().init();
-            Arc::<NodeDeath>::from(pinned)
-        };
+        let death = Arc::<NodeDeath>::from(
+            death
+                .pin_init_now::<core::convert::Infallible>(NodeDeath::new(
+                    info.node_ref.node.clone(),
+                    self.clone(),
+                    cookie,
+                ))
+                .unwrap_or_else(|(err, _)| match err {}),
+        );
```

Comment:
@wedsonaf can we refactor this code? So that we can merge lines 693-696
```rust
        let death = UniqueArc::try_new_uninit().map_err(|err| {
            thread.push_return_work(BR_ERROR);
            err
        })?;
```
with this?

Then I could remove `pin_init_now` and just use `pin_init` here.

Reply:
The problem with such a merge is that we'd have the memory allocation path (which may block for a while if the OOM killer is triggered) while holding a mutex. This is ok wrt safety, but is different from what the C version does.

It is unclear to me if the C version did this as some targeted optimisation after some profiling or if it's some potentially premature optimisation. I mimicked to C version here to reduce the risk of regressions.

Anyway, since we are unclear about the future of this codebase, I think it is ok to refactor for now. It would be nice if we had a long-term plan for supporting this scenario though.

Reply:
Oh I totally overlooked that `lock()` call. I think we should keep it this way.